---
title: "Filter DArT data with dartR"
output: html_document
---

This markdown provides a series of commands that may be useful when exploring and filtering DArT datasets  
It relies on the package `dartRverse` (using v. 1.0.6) and the packages:  
`dartR.base` (v. 1.0.8)  
`dartR.data` (v. 1.0.2)  
`dartR.spatial` (v. 0.78)  

Set the working directory where input files are  
Input files are:  
1) A CSV file from DArT containing SNPs scored as "1" or "0" on two lines (alleles)  
2) A CSV file of individual metadata (header: `id,pop,lat,lon`)  
```{r}
setwd("~/projects/Rhagodia/DArT")
```
Optionally, can remove outgroups from the start (in `meta.csv`) or  
can read in a `outgroup_samples.txt` file later with ids of outgroups to exclude  

Load library and designate input  
```{r}
suppressMessages(library("dartRverse"))
snps_file <- "Report_DRhag23-8571_SNP_1.csv"
meta_file <- "meta.csv"
```

Read in the DArT data (can take a bit for larger datasets)
```{r}
# mono.rm: remove monomorphic loci
# recalc: recalculate stats in case individuals were not included in the meta_file (dropped)
# probar: progress bar
mygenl <- gl.read.dart(snps_file, ind.metafile = meta_file,
	mono.rm = TRUE, recalc = TRUE, probar = TRUE)
```

Save a copy in case of shutdown/error or for convenience later
```{r}
gl.save(mygenl, file = "input.rdata")
```

If wanting to load later, start here
```{r}
mygenl <- gl.load("input.rdata")
```

Drop outgroups if desired (and if not done via `meta.csv`)
```{r}
outgroups <- read.table("outgroup_samples.txt", header = FALSE)[, 1]
mygenl <- gl.drop.ind(mygenl, outgroups, mono.rm = TRUE, recalc = TRUE)
cat(paste0("Filtered to a genlight of ", nInd(mygenl), " individuals and ",
	nLoc(mygenl), " SNPs\n"))
snps_remain <- nLoc(mygenl)
```

# Unfiltered data summary
Interactively map population locations (no output)
```{r}
gl.map.interactive(mygenl)
```

Assess various stats (and plot to file, if desired, by running the bounding statements below)
```{r}
pdf("unfiltered.pdf", width = 8, height = 12)
```
Read depth
```{r}
gl.report.rdepth(mygenl)
```
Tag length (sequence length)
```{r}
gl.report.taglength(mygenl)
```
Call rate (missing data)
```{r}
gl.report.callrate(mygenl, method = "loc")
gl.report.callrate(mygenl, method = "ind")
```
Reproducibility (whether calls are repeatable, averaged over both alleles)
```{r}
gl.report.reproducibility(mygenl)
```
Secondaries (multiple SNPs per tag; also gives an estimate of invariant loci)
```{r}
gl.report.secondaries(mygenl)
```
```{r}
invisible(dev.off())
```

Larger plots can be generated
```{r}
pdf("unfiltered_graphs.pdf", width = 21, height = 21)
```
Smear plot (can take a while with large datasets)
```{r}
smearplotgl <- gl.smearplot(mygenl, ind.labels = FALSE)
```
Minor allele frequency
```{r}
gl.report.maf(mygenl)
```
```{r}
invisible(dev.off())
```

# Filter data for various analyses
Change the name of the object for altering
```{r}
filtered_genl <- mygenl
snps_remain <- nLoc(filtered_genl)
```

Remove SNPs with low/high read depth (lower confidence in a call, higher error potentially)
```{r}
filtered_genl <- gl.filter.rdepth(filtered_genl, lower = 6, upper = 100)
diff <- snps_remain - nLoc(filtered_genl)
cat(paste0("Removed ", diff, " SNPs\n"))
cat(paste0("Filtered to a genlight of ", nInd(filtered_genl), " individuals and ",
	nLoc(filtered_genl), " SNPs\n"))
snps_remain <- nLoc(filtered_genl)
```

Filter based on tag length; remove short loci depending on distribution
```{r}
filtered_genl <- gl.filter.taglength(filtered_genl, lower = 60, upper = 69)
diff <- snps_remain - nLoc(filtered_genl)
cat(paste0("Removed ", diff, " SNPs\n"))
cat(paste0("Filtered to a genlight of ", nInd(filtered_genl), " individuals and ",
	nLoc(filtered_genl), " SNPs\n"))
snps_remain <- nLoc(filtered_genl)
```

Filter based on reproducibility (a proxy for problems)
```{r}
filtered_genl <- gl.filter.reproducibility(filtered_genl, threshold = 0.95)
diff <- snps_remain - nLoc(filtered_genl)
cat(paste0("Removed ", diff, " SNPs\n"))
cat(paste0("Filtered to a genlight of ", nInd(filtered_genl), " individuals and ",
	nLoc(filtered_genl), " SNPs\n"))
snps_remain <- nLoc(filtered_genl)
```

Remove SNPs found in less than a proportion of samples  
(do not use this here if wanting to estimate diversity stats per population later)
```{r}
filtered_genl <- gl.filter.callrate(filtered_genl, method = "loc", threshold = 0.95)
diff <- snps_remain - nLoc(filtered_genl)
cat(paste0("Removed ", diff, " SNPs\n"))
cat(paste0("Filtered to a genlight of ", nInd(filtered_genl), " individuals and ",
	nLoc(filtered_genl), " SNPs\n"))
snps_remain <- nLoc(filtered_genl)
```

Filter based on minor allele frequency  
(if a value > 1 is used, it is interpreted as minor allele count;  
this is preferable for a dataset of multiple populations)  
(do not use this if wanting to calculate diversity stats later)
```{r}
filtered_genl <- gl.filter.maf(filtered_genl, threshold = 3)
diff <- snps_remain - nLoc(filtered_genl)
cat(paste0("Removed ", diff, " SNPs\n"))
cat(paste0("Filtered to a genlight of ", nInd(filtered_genl), " individuals and ",
	nLoc(filtered_genl), " SNPs\n"))
snps_remain <- nLoc(filtered_genl)
```

Filter out individuals with large amounts of missing data (if present)
```{r}
filtered_genl <- gl.filter.callrate(filtered_genl, method = "ind", threshold = 0.50)
diff <- snps_remain - nLoc(filtered_genl)
cat(paste0("Removed ", diff, " SNPs\n"))
cat(paste0("Filtered to a genlight of ", nInd(filtered_genl), " individuals and ",
	nLoc(filtered_genl), " SNPs\n"))
snps_remain <- nLoc(filtered_genl)
```

Remove monomorphic and fully NA loci (if any are left)  
(do not use this if wanting to estimate diversity stats later)
```{r}
filtered_genl <- gl.filter.monomorphs(filtered_genl)
diff <- snps_remain - nLoc(filtered_genl)
cat(paste0("Removed ", diff, " SNPs\n"))
cat(paste0("Filtered to a genlight of ", nInd(filtered_genl), " individuals and ",
	nLoc(filtered_genl), " SNPs\n"))
snps_remain <- nLoc(filtered_genl)
```

If doing analyses sensitive to linkage (e.g. Structure), drop secondary SNPs  
Can also do this with `method = "random"`  
(do not use this if wanting to calculate autosomal diversity stats later)
```{r}
filtered_genl <- gl.filter.secondaries(filtered_genl, method = "best")
diff <- snps_remain - nLoc(filtered_genl)
cat(paste0("Removed ", diff, " SNPs\n"))
cat(paste0("Filtered to a genlight of ", nInd(filtered_genl), " individuals and ",
	nLoc(filtered_genl), " SNPs\n"))
snps_remain <- nLoc(filtered_genl)
```

To avoid linkage, use a Hamming filter to remove tags that are too similar (the same locus)  
The threshold (default 0.2) needs to consider the length of a tag (69 bp)  
Assume that if there are fewer than c. 8 bp differences, it is the same locus  
The threshold is then 8/69 = c. 0.12  
This can take a long time (not recommended if > 20 k SNPs remain; may have to run overnight)  
(do not use this if wanting to calculate diversity stats later)
```{r}
filtered_genl <- gl.filter.hamming(filtered_genl, tag.length = 69, rs = 5,
	threshold = 0.12, pb = TRUE)
diff <- snps_remain - nLoc(filtered_genl)
cat(paste0("Removed ", diff, " SNPs\n"))
cat(paste0("Filtered to a genlight of ", nInd(filtered_genl), " individuals and ",
	nLoc(filtered_genl), " SNPs\n"))
snps_remain <- nLoc(filtered_genl)
```

Save the object for import later with `gl.load`
```{r}
gl.save(filtered_genl, file = "filtered_genl.rdata")
```

If wanting to load later, start here
```{r}
filtered_genl <- gl.load("filtered_genl.rdata")
```

# Visualise the filtered data
Recalculate metrics in case samples have been dropped
```{r}
filtered_genl <- gl.recalc.metrics(filtered_genl)
```

## PCA
Run a PCA of the genlight SNPs
```{r}
pca <- gl.pcoa(filtered_genl, nfactors = 5, parallel = TRUE, n.cores = 8)
```

Start creating a PDF and then choose to run either/both plotting methods
```{r}
pdf("pca.pdf")
```

Use dartR inbuilt for plotting
```{r}
gl.pcoa.plot(pca, filtered_genl, pop.labels = "pop",
	xaxis = 1, yaxis = 2, zaxis = NULL, interactive = FALSE)
cat(paste0("Plotting a PCA based on ", snps_remain, " SNPs\n"))
```

Or use custom plotting (depends on pop being labeled with underscore before group, e.g. pop1_sp1)  
First set colour palette and points depending on number of groups (four or two here)
```{r}
# colour palette and point choices for number of groups/taxa
col_pal <- c("#d433da",
	"#0062ff",
	"#218a21",
	"#ff2e38"
)
point_ch <- c(22, 21, 24, 23)
```
```{r}
# colour palette and point choices for number of groups/taxa
col_pal <- c("#0062ff",
	"#218a21"
)
point_ch <- c(21, 24)
```
```{r}
# get groups
groups <- vector()
for (entry in filtered_genl$pop) {
	group <- strsplit(entry, "_")[[1]][2]
	groups <- c(groups, group)
}
# set colours and points
colours <- vector(length = nInd(filtered_genl))
points <- vector(length = nInd(filtered_genl))
index <- 1
for (group in unique(groups[order(groups)])) {
	indices <- which(groups == group)
	colours[indices] <- col_pal[index]
	points[indices] <- point_ch[index]
	index <- index + 1
}
# define a function to plot
plot_pca <- function(pca, numsnps, colours, points, x = 1, y = 2) {
	# grab elements
	pca_var <- round(100 * pca$eig / sum(pca$eig), digits = 1)
	method <- paste0("glPca (", numsnps, " SNPs)")
	pca_scores <- pca$scores
	# plot
	par(mar = c(5.1, 6.1, 5.1, 2.1), cex.lab = 2)
	plot(pca_scores[, x], pca_scores[, y],
		xlab = paste0("PC", x, " (", pca_var[x], "%)"),
		ylab = paste0("PC", y, " (", pca_var[y], "%)"),
		main = paste0("Principal Component Analysis\n", method),
		bg = colours,
		col = "black",
		lwd = 0.25,
		pch = points,
		asp = 1,
		cex = 2,
		panel.first = {
			grid()
			abline(h = 0, v = 0)
		},
		cex.axis = 1.5,
		cex.main = 1.5)
	par(mar = c(5.1, 4.1, 4.1, 2.1), cex.lab = 1)
}
```

Plot
```{r}
num_snps <- nLoc(filtered_genl)
plot_pca(pca, num_snps, colours, points, x = 1, y = 2)
plot_pca(pca, num_snps, colours, points, x = 1, y = 3)
plot_pca(pca, num_snps, colours, points, x = 1, y = 4)
plot_pca(pca, num_snps, colours, points, x = 2, y = 3)
plot_pca(pca, num_snps, colours, points, x = 2, y = 4)
plot_pca(pca, num_snps, colours, points, x = 3, y = 4)
```

Plot a legend
```{r}
groupdf <- data.frame(cbind(groups, colours, points))
leg_df <- unique(groupdf[order(groupdf[, 1]), ])
plot(1, type = "n", axes = FALSE, xlab = "", ylab = "")
legend("center", legend = leg_df[, 1],
	pch = as.numeric(leg_df[, 3]), pt.bg = leg_df[, 2],
	col = "black", title = NULL, cex = 1.5,
	pt.cex = 2, pt.lwd = 0.25)
```

```{r}
invisible(dev.off())
```

Or plot interactively
```{r}
gl.pcoa.plot(pca, filtered_genl, pop.labels = "pop", xaxis = 1, yaxis = 2, zaxis = NULL, interactive = TRUE)
```
Or plot interactively in 3D
```{r}
gl.pcoa.plot(pca, filtered_genl, pop.labels = "pop", xaxis = 1, yaxis = 2, zaxis = 3, interactive = TRUE)
```

# Output files for analysis
If relevant, generate a fasta of concatenated SNPs and corresponding distance matrix  
If wanting to run Structure, output a formatted input file  

Distances  
(method 3 = concatenate SNPs, with ambiguities for hets)
```{r}
library("ape")
library("pofadinr")
gl2fasta(filtered_genl, outfile = "filtered_SNPs.fasta", method = 3, outpath = ".")
mydnabin <- fasta2DNAbin("filtered_SNPs.fasta")
temp <- as.character(mydnabin)
temp[temp == "n"] <- "?"
mydnabin <- as.DNAbin(temp)
distance <- dist.snp(mydnabin, model = "GENPOFAD", as.matrix = TRUE)
outfile <- file("filtered_GENPOFAD_dist.txt", open = "w")
writeLines(as.character(nInd(filtered_genl)), con = outfile)
write.table(distance, file = outfile, sep = " ", append = TRUE,
	col.names = FALSE, quote = FALSE)
close(outfile)
```

Structure  
(first, replace pop with a number in the genlight)
```{r}
strgenl <- filtered_genl
index <- 1
pops <- sort(unique(strgenl$pop))
for (pop in pops) {
	levels(strgenl$pop)[match(pop, levels(strgenl$pop))] <- index
	index <- index + 1
}
gl2structure(strgenl, outfile = "filtered.str", add.columns = strgenl$pop,
	export.marker.names = FALSE, outpath = ".")
coding <- cbind(strgenl$ind.names, strgenl$pop)
write.table(coding[order(as.numeric(coding[, 2])), ], file = "filtered_str_pops.txt",
	row.names = FALSE, col.names = FALSE, quote = FALSE)
```

SNAPPER  
Load the filtered dataset (secondaries already removed), then downsample and remove monomorphic  
Filter for high coverage, then generate the output removing SNPs only variable in a single population
```{r}
snapper_genl <- gl.load("filtered_genl_all.rdata")
keep <- read.table("snapper/snapper_samples.txt", header = FALSE)[, 1]
snapper_genl <- gl.keep.ind(snapper_genl, as.character(keep), mono.rm = TRUE, recalc = TRUE)
snapper_genl <- gl.filter.callrate(snapper_genl, method = "loc", threshold = 1)
gl2snapper(snapper_genl, outfile = "snapper_snapp.nex", outpath = ".", rm.autapomorphies = TRUE, nloc = 1000)
```
Note: BEUAti may have issues reading this in, so you may have to change `integerdata` to `snp` in the output file  

NEXUS of binary characters  
This could be useful for SplitsTree and uncorrected P-distances  
As an example, use the `snapper_genl` just filtered above (or modify for `filtered_genl`)  
If interested in changing display labels, read in a text file with two columns (original and new), no header  
```{r}
mygenl <- gl.subsample.loci(snapper_genl, 1000, method = "random")
mymat <- as.matrix(mygenl)

# rename if desired (comment out if not)
names <- read.table("snapper/rename.txt", header = FALSE, sep = "\t")
names$V1 <- as.character(names$V1)
for (index in seq_len(nrow(mymat))) {
	rownames(mymat)[index] <- names$V2[match(rownames(mymat)[index], names$V1)]
}

# generate output
mymat[is.na(mymat)] <- "?"
mymat[mymat == "0"] <- "00"
mymat[mymat == "1"] <- "01"
mymat[mymat == "2"] <- "11"
matrix_block <- apply(mymat, 1, function(x) paste0(x, collapse = ""))
data_block <- paste0("BEGIN TAXA;\n\tDIMENSIONS NTAX=", nrow(mymat), ";\n\t",
	"TAXLABELS ", paste0(rownames(mymat), collapse = " "), ";\nEND;\n",
	"BEGIN CHARACTERS;\n\tDIMENSIONS NCHAR=", 2 * ncol(mymat), ";\n\t",
	"FORMAT\n\t\tDATATYPE=STANDARD\n\t\tMISSING=?\n\t\t",
	"LABELS=NO\n\t\tSYMBOLS=\"01\";\n\tMATRIX")
outfile <- file("snapper_splits.nex", open = "w")
writeLines("#NEXUS", con = outfile)
writeLines(data_block, con = outfile)
for (entry in matrix_block) {
	writeLines(paste0(c("\t\t", entry), collapse = ""), con = outfile)
}
writeLines("\t;\nEND;", con = outfile)
close(outfile)
```

# Popgen
Load the dataset (if not loaded yet and starting here)
```{r}
pop_genl <- gl.load("filtered_genl.rdata")
```

Or copy the object filtered before
```{r}
pop_genl <- filtered_genl
```

Check sample coverage
```{r}
summary(pop_genl$pop)
```

## Pairwise Fst and IBD
Drop individuals if necessary
```{r}
drops <- read.table("drop_samples_fst.txt", header = FALSE)[, 1]
pop_genl <- gl.drop.ind(pop_genl, drops, mono.rm = TRUE, recalc = TRUE)
cat(paste0("Filtered to a genlight of ", nInd(pop_genl), " individuals and ",
	nLoc(pop_genl), " SNPs\n"))
snps_remain <- nLoc(pop_genl)
```

Calculate pairwise Fst with StAMPP and write to file
```{r}
fsts <- gl.fst.pop(pop_genl, nboots = 1)
### export the pairwise Fst values to file
write.table(fsts, file = "popgen_StAMPP_Fst.txt",
	quote = FALSE, row.names = TRUE)
```

Run IBD if interested (will use STaMPP Fst by default)
```{r}
pdf("all_StAMPP_IBD.pdf", width = 7, height = 7)
gl.ibd(pop_genl)
invisible(dev.off())
```

Alternatively, subset to the populations of interest for IBD  
(then rerun the above with a different output name and input genl variable)
```{r}
drops <- read.table("drop_pops.txt", header = FALSE)[, 1]
pop_genl_subset <- gl.drop.pop(pop_genl, drops, mono.rm = TRUE, recalc = TRUE)
cat(paste0("Filtered to a genlight of ", nInd(pop_genl_subset), " individuals and ",
	nLoc(pop_genl_subset), " SNPs\n"))
snps_remain <- nLoc(pop_genl_subset)
```

## Autosomal/SNP diversity stats
Ensure that filtering prior to this is appropriate for autosomal measures, namely:
- secondaries are present  
- monomorphic loci are present  
- loci will be completely present (no missing) per population being considered  

Drop individuals
```{r}
drops <- read.table("drop_samples_diversity.txt", header = FALSE)[, 1]
pop_genl <- gl.drop.ind(pop_genl, drops, mono.rm = FALSE, recalc = TRUE)
cat(paste0("Filtered to a genlight of ", nInd(pop_genl), " individuals and ",
	nLoc(pop_genl), " SNPs\n"))
snps_remain <- nLoc(pop_genl)
```

Autosomal/SNP heterozygosity  
To ensure no missing data, each population needs to be filtered separately  
Only loci found in all samples in that population will be kept (zero missing data)  
Note: currently, the function will not report Fis for autosomal(!)
```{r}
stde <- function(x) {
	sqrt(var(x, na.rm = TRUE) / sum(!is.na(x)))
}

pops <- sort(unique(pop_genl$pop))
summary <- as.data.frame(matrix(ncol = 15, nrow = length(pops)))
colnames(summary) <- c("Samples", "Loci_auto", "Ho_auto", "Ho_auto_se", "He_auto", "He_auto_se",
	"Loci_SNPs", "Poly_SNPs", "Mono_SNPs", "Ho_SNP", "Ho_SNP_se", "He_SNP", "He_SNP_se", "Fis", "Fis_se")
rownames(summary) <- pops
for (pop in pops) {
	# subset and calculate autosomal stats
	sub_genl <- gl.keep.pop(pop_genl, pop, mono.rm = FALSE, recalc = TRUE)
	sub_genl <- gl.filter.callrate(sub_genl, method = "loc", threshold = 1)
	get_invariant <- gl.report.secondaries(sub_genl)
	stats <- gl.report.heterozygosity(sub_genl, n.invariant = get_invariant[7, 2])
	summary[pop, "Samples"] <- nInd(sub_genl)
	summary[pop, "Loci_auto"] <- stats[1, "n.Loc"]
	summary[pop, "Ho_auto"] <- stats[1, "Ho.adj"]
	summary[pop, "Ho_auto_se"] <- stats[1, "Ho.adjSE"]
	summary[pop, "He_auto"] <- stats[1, "He.adj"]
	summary[pop, "He_auto_se"] <- stats[1, "He.adjSE"]

	# calculate stats from the equivalent unlinked SNPs
	filt_sub_genl <- gl.filter.secondaries(sub_genl, method = "best")
	snp_stats <- gl.report.heterozygosity(filt_sub_genl, n.invariant = 0)
	summary[pop, "Loci_SNPs"] <- snp_stats[1, "n.Loc"]
	summary[pop, "Poly_SNPs"] <- snp_stats[1, "polyLoc"]
	summary[pop, "Mono_SNPs"] <- snp_stats[1, "monoLoc"]
	summary[pop, "Ho_SNP"] <- snp_stats[1, "Ho"]
	summary[pop, "Ho_SNP_se"] <- snp_stats[1, "HoSE"]
	summary[pop, "He_SNP"] <- snp_stats[1, "uHe"]
	summary[pop, "He_SNP_se"] <- snp_stats[1, "uHeSE"]

	# the dartR package has a bug that doesn't remove NA from standard error
	# filter out non-variant to address this and grab the Fis (doesn't change)
	filt_sub_genl <- gl.filter.monomorphs(filt_sub_genl)
	snp_stats <- gl.report.heterozygosity(filt_sub_genl, n.invariant = 0)
	summary[pop, "Fis"] <- snp_stats[1, "FIS"]
	summary[pop, "Fis_se"] <- snp_stats[1, "FISSE"]
}
```

Write to file
```{r}
write.table(summary, file = paste0("popgen_diversity_stats.txt"),
	quote = FALSE, row.names = TRUE)
```

Plot diversity stats to pdf
```{r}
mybarplot <- function(data_val, data_se, names, main) {
	myplot <- barplot(data_val, names = names, las = 2,
		main = main,
		ylim = c(min(c(0, 1.2 * (data_val - data_se * 2))),
			max(0, 1.2 * max(data_val + data_se * 2))))
	arrows(myplot, y0 = data_val + data_se * 2, y1 = data_val - data_se * 2,
		angle = 90, code = 3, length = 0.1)
}

pdf("popgen_diversity_stats.pdf", width = 10, height = 10)
mybarplot(summary$Ho_auto, summary$Ho_auto_se, names = rownames(summary),
	main = "Autosomal observed heterozygosity (Ho)")
mybarplot(summary$Ho_SNP, summary$Ho_SNP_se, names = rownames(summary),
	main = "SNP observed heterozygosity (Ho)")
mybarplot(summary$He_auto, summary$He_auto_se, names = rownames(summary),
	main = "Autosomal expected heterozygosity (He)")
mybarplot(summary$He_SNP, summary$He_SNP_se, names = rownames(summary),
	main = "SNP unbiased expected heterozygosity (uHe)")
mybarplot(summary$Fis, summary$Fis_se, names = rownames(summary),
		main = "Inbreeding coefficient Fis\n(1 - Ho / uHe)")
invisible(dev.off())
```
